dev/TeachingTip/TeachingTip.cpp

```diff
@@ -1340,249 +1387,340 @@ void TeachingTip::StartContractToClose()
+            availability[winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom] = false;
+        }
+        // If the tip is too wide to fit between the center of the target and the left edge of the window.
+        if (contentWidth - MinimumTipEdgeToTailCenter() > availableBoundsAroundTarget.Left + (targetBounds.Width / 2.0f))
+        {
+            availability[winrt::TeachingTipPlacementMode::TopEdgeAlignedLeft] = false;
+            availability[winrt::TeachingTipPlacementMode::BottomEdgeAlignedLeft] = false;
+        }
+        // If the tip is too wide to fit in the window when the tail is centered horizontally on the target and the tip.
+        if (contentWidth / 2.0f > availableBoundsAroundTarget.Left + (targetBounds.Width / 2.0f) ||
+            contentWidth / 2.0f > availableBoundsAroundTarget.Right + (targetBounds.Width / 2.0f))
+        {
+            availability[winrt::TeachingTipPlacementMode::Top] = false;
+            availability[winrt::TeachingTipPlacementMode::Bottom] = false;
+        }
+        // If the tip is too wide to fit between the center of the target and the right edge of the window.
+        if (contentWidth - MinimumTipEdgeToTailCenter() > availableBoundsAroundTarget.Right + (targetBounds.Width / 2.0f))
+        {
+            availability[winrt::TeachingTipPlacementMode::TopEdgeAlignedRight] = false;
+            availability[winrt::TeachingTipPlacementMode::BottomEdgeAlignedRight] = false;
+        }
+        // If the tip is too wide to fit between the right edge of the target and the right edge of the window.
+        if (tipWidth > availableBoundsAroundTarget.Right)
+        {
+            availability[winrt::TeachingTipPlacementMode::Right] = false;
+            availability[winrt::TeachingTipPlacementMode::RightEdgeAlignedTop] = false;
+            availability[winrt::TeachingTipPlacementMode::RightEdgeAlignedBottom] = false;
+        }
+
+        auto priorities = GetPlacementFallbackOrder(placement);
+
+        for (auto mode : priorities)
+        {
+            if (availability[mode])
             {
-                return winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom;
+                return std::make_tuple(mode, false);
             }
         }
     }
-    // The teaching tip wont fit anywhere, just return top.
-    return winrt::TeachingTipPlacementMode::Top;
+    // The teaching tip wont fit anywhere, set tipDoesNotFit to indicate that we should not open.
+    return std::make_tuple(winrt::TeachingTipPlacementMode::Top, true);
+}
+
+std::tuple<winrt::Rect, winrt::Thickness, winrt::Thickness> TeachingTip::DetermineSpaceAroundTarget()
+{
+    auto shouldConstrainToRootBounds = ShouldConstrainToRootBounds();
+    // We use the current window/XamlRoot bounds to determine how much space is between the bottom and left edge of the target
+    // and the bottom and left edge of the window/XamlRoot.
+    auto windowBounds = m_testWindowBounds;
```

Comment:
This talks about XamlRoot, but where is XamlRoot used? CoreWindow.Bounds will be 0x0 in islands scenarios so we should use XamlRoot if it's available.